kv: TestTxnCoordSenderHeartbeat failed under stress

SHA: https://github.com/cockroachdb/cockroach/commits/eb028e6e962b864b79c9d0c610d2c44b26a7d0c4

Parameters:
```
TAGS=
GOFLAGS=-race
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=793325&tab=buildLog

```
=== RUN   TestTxnCoordSenderHeartbeat
I180724 06:00:55.145508 1830 gossip/gossip.go:363  [n1] NodeDescriptor set to node_id:1 address:<network_field:"" address_field:"" > attrs:<> locality:<> ServerVersion:<major_val:0 minor_val:0 patch:0 unstable:0 > 
I180724 06:00:55.152978 1830 storage/replica_command.go:279  [n1,s1,r1/1:/M{in-ax}] initiating a split of this range at key "b" [r2]
--- FAIL: TestTxnCoordSenderHeartbeat (0.28s)
=== RUN   TestTxnCoordSenderHeartbeat/pusher:"a"
==================
WARNING: DATA RACE
Write at 0x00c4206142c0 by goroutine 207:
  github.com/cockroachdb/cockroach/pkg/kv.TestTxnCoordSenderHeartbeat.func1()
      /go/src/github.com/cockroachdb/cockroach/pkg/kv/txn_coord_sender_test.go:424 +0x11e
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:777 +0x16d

Previous read at 0x00c4206142c0 by goroutine 96:
  github.com/cockroachdb/cockroach/pkg/kv.(*TxnCoordSender).heartbeatLoop()
      /go/src/github.com/cockroachdb/cockroach/pkg/kv/txn_coord_sender.go:760 +0x7b
  github.com/cockroachdb/cockroach/pkg/kv.(*TxnCoordSender).startHeartbeatLoopLocked.func1()
      /go/src/github.com/cockroachdb/cockroach/pkg/kv/txn_coord_sender.go:940 +0x53
  github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTask.func1()
      /go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:266 +0xf3

Goroutine 207 (running) created at:
  testing.(*T).Run()
      /usr/local/go/src/testing/testing.go:824 +0x564
  github.com/cockroachdb/cockroach/pkg/kv.TestTxnCoordSenderHeartbeat()
      /go/src/github.com/cockroachdb/cockroach/pkg/kv/txn_coord_sender_test.go:421 +0x42f
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:777 +0x16d

Goroutine 96 (running) created at:
  github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTask()
      /go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:261 +0x14c
  github.com/cockroachdb/cockroach/pkg/kv.(*TxnCoordSender).startHeartbeatLoopLocked()
      /go/src/github.com/cockroachdb/cockroach/pkg/kv/txn_coord_sender.go:938 +0x358
  github.com/cockroachdb/cockroach/pkg/kv.(*TxnCoordSender).Send()
      /go/src/github.com/cockroachdb/cockroach/pkg/kv/txn_coord_sender.go:562 +0x1128
  github.com/cockroachdb/cockroach/pkg/internal/client.(*DB).sendUsingSender()
      /go/src/github.com/cockroachdb/cockroach/pkg/internal/client/db.go:617 +0x123
  github.com/cockroachdb/cockroach/pkg/internal/client.(*Txn).Send()
      /go/src/github.com/cockroachdb/cockroach/pkg/internal/client/txn.go:1069 +0x6b8
  github.com/cockroachdb/cockroach/pkg/internal/client.(*Txn).Send-fm()
      /go/src/github.com/cockroachdb/cockroach/pkg/internal/client/txn.go:569 +0xa3
  github.com/cockroachdb/cockroach/pkg/internal/client.sendAndFill()
      /go/src/github.com/cockroachdb/cockroach/pkg/internal/client/db.go:542 +0x151
  github.com/cockroachdb/cockroach/pkg/internal/client.(*Txn).Run()
      /go/src/github.com/cockroachdb/cockroach/pkg/internal/client/txn.go:569 +0xfd
  github.com/cockroachdb/cockroach/pkg/storage.(*Replica).adminSplitWithDescriptor.func1()
      /go/src/github.com/cockroachdb/cockroach/pkg/storage/replica_command.go:309 +0x345
  github.com/cockroachdb/cockroach/pkg/internal/client.(*DB).Txn.func1()
      /go/src/github.com/cockroachdb/cockroach/pkg/internal/client/db.go:580 +0x5e
  github.com/cockroachdb/cockroach/pkg/internal/client.(*Txn).exec()
      /go/src/github.com/cockroachdb/cockroach/pkg/internal/client/txn.go:831 +0xf6
  github.com/cockroachdb/cockroach/pkg/internal/client.(*DB).Txn()
      /go/src/github.com/cockroachdb/cockroach/pkg/internal/client/db.go:579 +0x13b
  github.com/cockroachdb/cockroach/pkg/storage.(*Replica).adminSplitWithDescriptor()
      /go/src/github.com/cockroachdb/cockroach/pkg/storage/replica_command.go:282 +0xc82
  github.com/cockroachdb/cockroach/pkg/storage.(*Replica).AdminSplit()
      /go/src/github.com/cockroachdb/cockroach/pkg/storage/replica_command.go:159 +0x29f
  github.com/cockroachdb/cockroach/pkg/storage.(*Replica).executeAdminBatch()
      /go/src/github.com/cockroachdb/cockroach/pkg/storage/replica.go:2606 +0xf9d
  github.com/cockroachdb/cockroach/pkg/storage.(*Replica).sendWithRangeID()
      /go/src/github.com/cockroachdb/cockroach/pkg/storage/replica.go:2000 +0x77f
  github.com/cockroachdb/cockroach/pkg/storage.(*Replica).Send()
      /go/src/github.com/cockroachdb/cockroach/pkg/storage/replica.go:1942 +0xb5
  github.com/cockroachdb/cockroach/pkg/storage.(*Store).Send()
      /go/src/github.com/cockroachdb/cockroach/pkg/storage/store.go:3004 +0x73b
  github.com/cockroachdb/cockroach/pkg/storage.(*Stores).Send()
      /go/src/github.com/cockroachdb/cockroach/pkg/storage/stores.go:185 +0x133
  github.com/cockroachdb/cockroach/pkg/kv.(*senderTransport).SendNext()
      /go/src/github.com/cockroachdb/cockroach/pkg/kv/transport.go:383 +0x2b0
  github.com/cockroachdb/cockroach/pkg/kv.(*localTestClusterTransport).SendNext()
      /go/src/github.com/cockroachdb/cockroach/pkg/kv/local_test_cluster_util.go:45 +0x8a
  github.com/cockroachdb/cockroach/pkg/kv.(*DistSender).sendToReplicas()
      /go/src/github.com/cockroachdb/cockroach/pkg/kv/dist_sender.go:1257 +0x5a7
  github.com/cockroachdb/cockroach/pkg/kv.(*DistSender).sendRPC()
      /go/src/github.com/cockroachdb/cockroach/pkg/kv/dist_sender.go:367 +0x2cb
  github.com/cockroachdb/cockroach/pkg/kv.(*DistSender).sendSingleRange()
      /go/src/github.com/cockroachdb/cockroach/pkg/kv/dist_sender.go:434 +0x272
  github.com/cockroachdb/cockroach/pkg/kv.(*DistSender).sendPartialBatch()
      /go/src/github.com/cockroachdb/cockroach/pkg/kv/dist_sender.go:1037 +0x3da
  github.com/cockroachdb/cockroach/pkg/kv.(*DistSender).divideAndSendBatchToRanges()
      /go/src/github.com/cockroachdb/cockroach/pkg/kv/dist_sender.go:715 +0x17d4
  github.com/cockroachdb/cockroach/pkg/kv.(*DistSender).Send()
      /go/src/github.com/cockroachdb/cockroach/pkg/kv/dist_sender.go:629 +0x59f
  github.com/cockroachdb/cockroach/pkg/internal/client.(*CrossRangeTxnWrapperSender).Send()
      /go/src/github.com/cockroachdb/cockroach/pkg/internal/client/db.go:213 +0xe3
  github.com/cockroachdb/cockroach/pkg/internal/client.(*DB).sendUsingSender()
      /go/src/github.com/cockroachdb/cockroach/pkg/internal/client/db.go:617 +0x123
  github.com/cockroachdb/cockroach/pkg/internal/client.(*DB).send()
      /go/src/github.com/cockroachdb/cockroach/pkg/internal/client/db.go:599 +0xb4
  github.com/cockroachdb/cockroach/pkg/internal/client.(*DB).(github.com/cockroachdb/cockroach/pkg/internal/client.send)-fm()
      /go/src/github.com/cockroachdb/cockroach/pkg/internal/client/db.go:565 +0xa3
  github.com/cockroachdb/cockroach/pkg/internal/client.sendAndFill()
      /go/src/github.com/cockroachdb/cockroach/pkg/internal/client/db.go:542 +0x151
  github.com/cockroachdb/cockroach/pkg/internal/client.(*DB).Run()
      /go/src/github.com/cockroachdb/cockroach/pkg/internal/client/db.go:565 +0xed
  github.com/cockroachdb/cockroach/pkg/internal/client.(*DB).AdminSplit()
      /go/src/github.com/cockroachdb/cockroach/pkg/internal/client/db.go:471 +0xe2
  github.com/cockroachdb/cockroach/pkg/kv.TestTxnCoordSenderHeartbeat()
      /go/src/github.com/cockroachdb/cockroach/pkg/kv/txn_coord_sender_test.go:407 +0x2a8
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:777 +0x16d
==================
    --- FAIL: TestTxnCoordSenderHeartbeat/pusher:"a" (0.08s)
    	testing.go:730: race detected during execution of test

```

--- a/pkg/kv/local_test_cluster_util.go
+++ b/pkg/kv/local_test_cluster_util.go
@@ -57,10 +57,32 @@ func InitFactoryForLocalTestCluster(
 	stopper *stop.Stopper,
 	gossip *gossip.Gossip,
 ) client.TxnSenderFactory {
+	return NewTxnCoordSenderFactory(
+		TxnCoordSenderFactoryConfig{
+			AmbientCtx: log.AmbientContext{Tracer: st.Tracer},
+			Settings:   st,
+			Clock:      clock,
+			Stopper:    stopper,
+		},
+		NewDistSenderForLocalTestCluster(st, nodeDesc, tracer, clock, latency, stores, stopper, gossip),
+	)
+}
+
+// NewDistSenderForLocalTestCluster creates a DistSender for a LocalTestCluster.
+func NewDistSenderForLocalTestCluster(
+	st *cluster.Settings,
+	nodeDesc *roachpb.NodeDescriptor,
+	tracer opentracing.Tracer,
+	clock *hlc.Clock,
+	latency time.Duration,
+	stores client.Sender,
+	stopper *stop.Stopper,
+	gossip *gossip.Gossip,
+) *DistSender {
 	retryOpts := base.DefaultRetryOptions()
 	retryOpts.Closer = stopper.ShouldQuiesce()
 	senderTransportFactory := SenderTransportFactory(tracer, stores)
-	distSender := NewDistSender(DistSenderConfig{
+	return NewDistSender(DistSenderConfig{
 		AmbientCtx:      log.AmbientContext{Tracer: st.Tracer},
 		Settings:        st,
 		Clock:           clock,
@@ -81,14 +103,4 @@ func InitFactoryForLocalTestCluster(
 			},
 		},
 	}, gossip)
-
-	return NewTxnCoordSenderFactory(
-		TxnCoordSenderFactoryConfig{
-			AmbientCtx: log.AmbientContext{Tracer: st.Tracer},
-			Settings:   st,
-			Clock:      clock,
-			Stopper:    stopper,
-		},
-		distSender,
-	)
 }

--- a/pkg/kv/txn_coord_sender_test.go
+++ b/pkg/kv/txn_coord_sender_test.go
@@ -409,6 +409,24 @@ func TestTxnCoordSenderHeartbeat(t *testing.T) {
 		t.Fatal(err)
 	}
 
+	// Make a db with a short heartbeat interval.
+	ambient := log.AmbientContext{Tracer: tracing.NewTracer()}
+	tsf := NewTxnCoordSenderFactory(
+		TxnCoordSenderFactoryConfig{
+			AmbientCtx: ambient,
+			// Short heartbeat interval.
+			HeartbeatInterval: time.Millisecond,
+			Settings:          s.Cfg.Settings,
+			Clock:             s.Clock,
+			Stopper:           s.Stopper,
+		},
+		NewDistSenderForLocalTestCluster(
+			s.Cfg.Settings, &roachpb.NodeDescriptor{NodeID: 1},
+			ambient.Tracer, s.Clock, s.Latency, s.Stores, s.Stopper, s.Gossip,
+		),
+	)
+	quickHeartbeatDB := client.NewDB(ambient, tsf, s.Clock)
+
 	// We're going to test twice. In both cases the heartbeat is supposed to
 	// notice that its transaction is aborted, but:
 	// - once the abort span is populated on the txn's range.
@@ -420,9 +438,9 @@ func TestTxnCoordSenderHeartbeat(t *testing.T) {
 	// abort span.
 	for _, pusherKey := range []roachpb.Key{keyA, keyC} {
 		t.Run(fmt.Sprintf("pusher:%s", pusherKey), func(t *testing.T) {
-			initialTxn := client.NewTxn(s.DB, 0 /* gatewayNodeID */, client.RootTxn)
+			// Make a db with a short heartbeat interval.
+			initialTxn := client.NewTxn(quickHeartbeatDB, 0 /* gatewayNodeID */, client.RootTxn)
 			tc := initialTxn.Sender().(*TxnCoordSender)
-			tc.TxnCoordSenderFactory.heartbeatInterval = 1 * time.Millisecond
 
 			if err := initialTxn.Put(ctx, keyA, []byte("value")); err != nil {
 				t.Fatal(err)
@@ -440,10 +458,7 @@ func TestTxnCoordSenderHeartbeat(t *testing.T) {
 						t.Fatal(pErr)
 					}
 					// Advance clock by 1ns.
-					// Locking the TxnCoordSender to prevent a data race.
-					tc.mu.Lock()
 					s.Manual.Increment(1)
-					tc.mu.Unlock()
 					if lastActive := txn.LastActive(); heartbeatTS.Less(lastActive) {
 						heartbeatTS = lastActive
 						return nil

